Ticket/359/maximum/projection

danielsf opened the pull request:
This PR adds a module that would create a maximum projection image more in line with Leonard and John's legacy code.

We still need to decide if this should be run as an independent module, or should be functionality called by the Suite2P motion correction pipeline. We probably should not merge this PR until that conversation is resolved. That being said, I think the code itself is ready for review.

aamster commented on src/ophys_etl/modules/maximum_projection/__main__.py:
```diff
@@ -0,0 +1,113 @@
+
+    The image is produced by
+    1) Downsampling the video to a specified frame rate
+    2) Applying a median filter to every frame in teh downsampled video
```
"the"

aamster commented on src/ophys_etl/modules/maximum_projection/utils.py:
```diff
@@ -0,0 +1,294 @@
+import scipy.ndimage as scipy_ndimage
+
+
+def scale_to_uint8(img: np.ndarray) -> np.ndarray:
```
Is this function not somewhere else in the codebase?

danielsf replied:
I don't think so. We definitely implemented it as a part of `staging/segmentation_dev`. I think it's an open question whether and how much of that branch ever gets merged into `main`.

I'll put this somewhere more generally accessible.

danielsf replied:
Actually: never mind.

`utils/array_utils/normalize_array`

danielsf replied:
There was apparently also already a `downsample_array` method. I've changed the code to use that, rather than my custom-implemented `decimate_video` method.

aamster commented on src/ophys_etl/modules/maximum_projection/utils.py:
```diff
@@ -0,0 +1,294 @@
+import pathlib
```
These seem like more than utils to me. What are your thoughts on creating a `MaximumProjection` class, and adding a `@classmethod` `from_path` to instantiate it?

danielsf replied:
Is this more than an aesthetic choice?

From my perspective, the output is just a numpy array, so we don't need to define a special output class, and there isn't a state that needs to be carried around during processing. I see the simplicity in having `FilteredMaximumProjection.from_array()` and `Filtered MaximumProjection.from_path()`. It seems overly complicated to me to have those methods return a numpy array, rather than some more complicated object.

aamster replied:
If not a class, then I would move the logic out of `utils` and make the functions private which are local to the module

aamster commented on src/ophys_etl/modules/maximum_projection/utils.py:
```diff
@@ -0,0 +1,294 @@
+
+    for i_out, i0 in enumerate(range(0, ntime_in, frames_to_group)):
+        i1 = min(ntime_in, i0+frames_to_group)
+        decimated_video[i_out, :, :] = np.mean(video[i0:i1, :, :], axis=0)
```
FYI the extra `:, :` are not needed

aamster reviewed (commented):
Overall good. I can't comment on the correctness of the algorithm, but it looks good from the example you produced.

My only suggestion is to create a `MaximumProjection` class for encapsulating logic to create the max projection instead of putting it in a `utils` module.

njmei commented on src/ophys_etl/modules/maximum_projection/__main__.py:
```diff
@@ -0,0 +1,113 @@
+    scale_to_uint8)
+
+
+class MaximumProjectionSchema(argschema.ArgSchema):
```
To avoid future confusion it might be good to call this whole module and the output image a `median filtered max projection`. `FilteredMaximumProjectionSchema`

danielsf replied:
Good point. Especially since we haven't yet figured out whether we actually want to do this automatically or when (i.e. as a part of motion correction or as a stand alone module).

danielsf commented:
@aamster @njmei 
I think this is ready for another look.

1) I renamed the module `median_filtered_max_projection`
2) I renamed the methods `median_filtered_max_projection_from_*` to make it a little clearer what was going on
3) @aamster I left the image generation methods in `utils.py`. The only better name I could think of for that script was `median_filtered_max_projeciton.py` and `from ophys_etl.modules.median_filtered_max_projection.median_filtered_max_projection import *` seemed too long. I think the fact that these utils are in `median_filtered_max_projection` should make it clear enough what is going on.
4) I altered the code to rely on the pre-existing methods for scaling an array to uint8 and downsampling a video. In doing so, I had to alter the logic of those methods slightly to make sure they did what I needed them to do. I only broke one test, but I think that test was wrong anyway (it was using `np.floor()` rather than `np.round()` when converting between a float and uint8).

Of all these, I think (4) is the most controversial, since I ended up touching code I didn't originally intend to.

aamster commented on src/ophys_etl/modules/median_filtered_max_projection/utils.py:
```diff
@@ -0,0 +1,213 @@
+    n_frames_from_hz)
+
+
+def apply_median_filter_to_video(
```
This appears private to this module. Could you please mark it as private?

danielsf replied:
I don't see it as private. For the moment it is only used here, but it is an operation ("apply a median filter to every frame in a video") that I could imagine us wanting to use in future data processing pipelines. I'm going to leave it as-is.
